site-src/geps/gep-1364.md

```diff
@@ -0,0 +1,328 @@
+plane.
+
+So, All Conditions Positive pros:
+* We're close already. Most conditions in the API are currently positive polarity.
+* Easier to understand - there are no double negatives. "Good: true" is less
+cognitive overhead than "NotGood: false".
+
+Cons:
+* Reduces flexibility - it can surprisingly difficult to avoid double negatives for
+conditions that describe error states, as in general programmers are more used
+to reporting "something went wrong" than they are "everything's okay".
+
+Not sure if pro or con:
+* Leans the design towards favoring conditions always being present, since you
+can't be sure if everything is good unless you see `AllGood: true`. The absence
+of a positive-polarity condition implies that the condition could be false. This
+puts this option more in line with the API guidelines on this point.
+
+#### Some Conditions Positive
+
+In this case, only a limited set of summary conditions are positive, and the rest
+are negative.
+
+Pros:
+* Error states can be described with `Error: true` instead of `NoError: false`.
+* Negative polarity error conditions are more friendly to not being present (since
+absence of `Error: true` implies everything's okay).
+
+Cons:
+* Double negatives in the happy case: `NoError: false`
+
+#### Decision
+
+Gateway API resource conditions will all be positive polarity.
+
+### Should conditions always be added?
+
+Since we are mandating that conditions should always be positive, we should also
+specify that controllers always place all conditions on each resource when they
+reconcile it, even if that update sets the status to `unknown` because of
+information not being available.
+
+This is becasue, whatever the API guidelines say, the absence of a positive
+condition means that a user can't be _sure_ that the condition is okay. So, we
+should ensure that all the positive conditions are always present, even if their
+status is `unknown`.
+
+### New and Updated Conditions
+
+#### `Attached`
+
+This GEP proposes replacing all conditions that indicate syntactic and semantic
+validity with one, `Attached` condition type, with the exception of the
+GatewayClass resource (because it's the root of the resource tree and doesn't
+attach to anything).
+
+That is, the proposal is to replace:
+* `Scheduled` on Gateway
+* `Detached` on Listener
+* `Accpeted` on Route
+
+with `Attached` in all these locations.
+
+GatewayClass will maintain the `Accepted` condition.
+
+All of these conditions share the following meanings:
+* The resource has been accepted for processing by the controller
+* The resource is syntatically and semantically valid, and internally consistent
+* The resource fits into a larger system of Gateway API resources, and there is
+is no missing information, including but not limited to:
+  * Any mandatory references resolve to existing resources (examples here are the
+  Gateway's gatewayClass field, or the `parentRefs` field in Route resources)
+  * Any specified TLS secrets exist
```

Comment:
Currently the API supports partially-valid resources. I believe this is proposing a partially-valid resource would be `Attached=false`, but the config is actually 'active' (or, I would describe it, "attached").

To me this feels a bit odd. The old API sort of had this issue as well.

Thinking purely from a pedantic-correctness POV (which may not be ideal - sometimes simpler is better), I would expect there to be two conditions (I use bad names intentionally to avoid confusion): "ThisResourceWasApplied" - any part of the rule had some effect on the system, and "ThisRuleIsEntirelyValid" - no errors found in the rule at all.

Its possible my concern would be addressed with a different word than "Attached", rather than more conditions, as well

WDYT?

Reply:
I guess this is somewhat addressed below.. but I am not sure its accurate? It seems we are defining a very small set of "failures that do not become unattached", but there are a lot more partially-accepted resources I think?

Reply:
The guideline I've been basing the "does something become unattached?" on is "Does this produce at least some configuration in the underlying data plane?". So I guess it sounds like I need to be more explicit with that guideline, but that aside from that, we're largely in agreement?

Reply:
I think it makes sense to have "Is this attached" and "is this 100% valid" as two separate things. 

Part of my confusion, I think, is I am not actually sure which fields result in "not attached". It seems _most_ issues are partially accepted. Which is fine - `Attached` just becomes basically a pretty high level acknowledgement that the controller is handling the resource

Reply:
Yeah I think I agree with @howardjohn's point here: "does this resource produce config" is distinct from "does this resource produce _valid_ config". I'm sure some implementations don't allow invalid (e.g. duplicate route names or something) to be sent to the data plane, but it doesn't feel like that's a requirement for the Gateway API at this juncture.

Reply:
I guess I wrote this pooly - by "Does this resource produce config" I meant to read "Is this resource valid enough to produce some config when combined with the rest of the resource set".

The thing I'm thinking of particularly here is that for HTTPRoute, invalid references still produce data path config because the match will respond with 500s. Whereas for TCPRoute etc, if you don't have a service to route to, there's no way you can build a forwarding path - there's nothing to forward to! So that won't produce any data path config at all, and so, in the current design, would cause the resource to fail to be attached.

I think that partial validity is an enormous minefield of weirdness, just like not applying config if it would break things.

Reply:
Is it possible for something like an HTTPRoute to define two paths, one of which is "valid enough" config, while the other is garbage?

If so, are there expectations or requirements from a conformance PoV ("always apply all or none") that we need to specify?

Reply:
> Is it possible for something like an HTTPRoute to define two paths, one of which is "valid enough" config, while the other is garbage?

Most cases of an invalid configuration in one route rule (with others being completely valid within the same HTTPRoute) seem to be misconfigurable only in the context of a reference to another resource and fall into the `ResolvedRefs` bucket (which seems stated below and possibly agreed upon that `Attached: true` be set with `ResolvedRefs: false` and return HTTP code 500 for the invalid rule).

One such example not covered is if you were to use a `HTTPRequestRedirectFilter` and `HTTPURLRewrite` in the list of filters for an individual `HTTPRouteRule` ([ref](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRequestRedirectFilter)). At the moment it doesn't appear we have prescribed any status code or other error handling for this error. This seems to fall into the bucket of things that cannot produce data plane configuration.

Going by the statements below, if this rule were the only one present `Attached: false` should be set, since it cannot result in any configuration (with another more specific error condition). If present with other valid rules, `Attached: true` should be set (with another more specific error condition again). I can definitely see the argument for setting `Attached: false` however in the latter case, since stated above, the `Attached` condition means `The resource is syntatically and semantically valid, and internally consistent`.

It seems that the `Valid: true` condition mentioned above ends up being somewhat equivalent to having  for example for HTTPRoute `ResolvedRefs: true` and no error conditions (and vice versa, `Valid: false` equivalent to `ResolvedRefs: false` *or* some more specific error condition). Maybe this framing helps a little with determining possible merits of having another condition?

> If so, are there expectations or requirements from a conformance PoV ("always apply all or none") that we need to specify?

Just doing a quick pass, I don't *think* I saw any existing conformance tests that cover this kind of thing so I agree that would be good to add.

Reply:
Avoiding "always apply all or none" is the reason this is tricky. There are mixed expectations about what happens if _some_ config is present but bad, and I'm attempting to thread the needle here and make this as user-friendly as possible while not becoming really difficult to implement or understand.

Reply:
I added a section on partial validity to hopefully make this clearer.